### PR TITLE
Add discriminator validation feature

### PR DESCRIPTION
**Pull request by gigaga**
From this PR, we are able to validate examples that contain "discriminator".
However to can validate a specification, inheritance have to be in the following form : 

```
---
openapi: 3.0.2
info:
  title: petstore Inheritance
  version: 1.0.7
paths:
  "/pet":
    post:
      tags:
      - pet
      summary: Add a new pet to the store
      description: Add a new pet to the store
      operationId: addPet
      requestBody:
        description: Create a new pet in the store
        content:
          application/json:
            schema:
              $ref: "#/components/schemas/Pet"
            examples:
              dog:
                value:
                  name: "12"
                  catName: "123"
                  type: "CAT"
        required: true 
      responses:
        '200':
          description: Successful operation
          content:
            application/json:
              schema:
                "$ref": "#/components/schemas/Pet"
        '405':
          description: Invalid input
components:
  schemas:
    Dog: 
      type: object
      properties:
        dogName:
          type: string
        type:
          type: string
          enum:
            - DOG
    Cat:
      type: object
      properties:
        catName:
          type: string
        type:
          type: string
          enum:
            - CAT
    Pet:
      required:
      - name
      - type
      type: object
      properties:  
        id:
          type: integer
          format: int64
          example: 10
        name:
          type: string
          example: doggie
        type:
          type: string
      discriminator:
        propertyName: type
      oneOf:
        - $ref: "#/components/schemas/Cat"
        - $ref: "#/components/schemas/Dog"
```

In other terms : 
- `mapping` keyword hasn't be present
- main schema has reference sub-schemas by using `oneOf` keyword, and not by using `anyOf` from sub-schemas
- each sub-schema has to specify the discriminator property with its related value

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -69,19 +82,42 @@
...
-    "json-pointer": "0.6.1",
-    "json-schema-ref-parser": "^9.0.9",
-    "jsonpath-plus": "^6.0.1",
-    "lodash.clonedeep": "^4.5.0",
-    "lodash.flatmap": "^4.5.0",
-    "lodash.flatten": "^4.4.0",
-    "lodash.merge": "^4.6.2",
-    "yaml": "^1.10.2"
-  }
+  "engines": {
```
Comment (codekie) [inline package.json:85]: Generally, don't modify the structure of the `package.json` unless it is really necessary for the feature.

If only new dependencies are required, only add these.

File: package.json [hunk 2]
```diff
@@ -1,50 +1,63 @@
...
   "bugs": {
     "url": "https://github.com/codekie/openapi-examples-validator/issues"
   },
-  "homepage": "https://github.com/codekie/openapi-examples-validator#readme",
+  "bundleDependencies": false,
+  "dependencies": {
+    "ajv": "^8.8.2",
+    "ajv-draft-04": "^1.0.0",
+    "ajv-oai": "1.2.1",
+    "ajv-openapi": "^2.0.0",
```
Comment (codekie) [inline package.json:46]: Why has this dependency been added, when it's not used?

Also, what are the benefits of this library over `ajv-oai` that already is in use?

**Review by codekie — CHANGES_REQUESTED** [review]
Quite some tests are failing yet (please check by running the following command):

```sh
npm run test
```

Some of the addressed issues already are pointed out when you either run the `build`- or `test`-task.

Also, every new feature requires its own set of automated tests, before it's going to be integrated.

File: package.json [hunk 3]
```diff
@@ -1,50 +1,63 @@
 {
-  "name": "openapi-examples-validator",
-  "version": "4.5.0",
-  "description": "Validates embedded examples in OpenAPI-JSONs",
-  "main": "dist/index.js",
-  "engines": {
-    "node": ">=8"
+  "_from": "openapi-examples-validator@4.5.0",
```
[the remark this answers is not in the document]
Comment (gigaga) [reply package.json:2]: So, sorry, it's a mistake :(

File: package.json [hunk 4]
```diff
@@ -1,50 +1,63 @@
...
-    "validate",
-    "examples"
-  ],
-  "author": "Josua Amann",
-  "license": "MIT",
   "bugs": {
     "url": "https://github.com/codekie/openapi-examples-validator/issues"
   },
-  "homepage": "https://github.com/codekie/openapi-examples-validator#readme",
+  "bundleDependencies": false,
```
[the remark this answers is not in the document]
Comment (gigaga) [reply package.json:41]: Mistake too :(

**Comment by gigaga** [comment]
2 unit-tests are still in error, but I don't know why :(

```
2) Main-module, for v3 should
       be able to handle `nullable`-types
         with nullable property, but no type set:

      AssertionError: expected false to equal true
      + expected - actual

      -false
      +true
``` 
==> probably because nullable: true if the default behavior for now


```
1) CLI-module
       ignore datatype formats
         should show no error when formats have been passed, with newline separated:
     Error: Command failed: node D:\Developpement\openapi-examples-validator\src\cli.js --ignore-formats "country-code-2
continental-status
license-plate" -- D:\Developpement\openapi-examples-validator\test\data\v3\unknown-formats.json
error: missing required argument 'filepath'

      at ChildProcess.exithandler (child_process.js:308:12)
      at maybeClose (internal/child_process.js:1048:16)
      at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5)
``` 
==> nothing was changed for that, so :(

File: test/specs/cli.js [hunk 5]
```diff
@@ -86,8 +86,8 @@ describe('CLI-module', function() {
                     .catch(({ stdout, stderr }) => {
                         stdout.should.contain('Errors found.', 'Stdout did not contain "Errors found." message');
                         stderr.should.contain(
-                            'unknown format \\"timestamp\\" is used in schema at '
-                            + 'path \\"#/properties/versions/items/properties/occurred',
+                            'unknown format "timestamp" ignored in schema at '
```
Comment (codekie) [inline test/specs/cli.js:89]: The new version behaves differently here.

The old version pointed out that there is an unknown format in use and then stopped the validation with an error, while the new version only drops a warning and proceeds with the validation. This also means that if there is a file with no other validation-errors, this warning (not error) may go unnoticed.

So for this particular case, we either have to restore the original behaviour...

.. or even better, provide a CLI-argument to force an error for unknown formats. This opens the opportunity to validate files with unknown formats, which isn't possible yet.

**Review by codekie — CHANGES_REQUESTED** [review]

**Comment by codekie** [comment]
The failing test is because the new version of `ajv` requires the `type` attribute to be set, if `nullable` is set. That was not a requirement in the older version.

I also have seen that there were other breaking changes, like:

- `dataPath` property replaced with `instancePath`
- "should" replaced with "must" in the messages
- property name is removed from "propertyName" keyword error message (it is still available in `error.params.propertyName`).
- The `exclusiveMinimum`-keyword has changed to `minimum`
- Unknown formats for types ([see comment](https://github.com/codekie/openapi-examples-validator/pull/156/files#diff-0ad62bc62fcbe89f735daedef0d8b3e28f410404902056b7ab83d3c6276d8147R89)) only log a warning, instead of throwing an error

.. as some have been pointed out in https://github.com/ajv-validator/ajv/releases/tag/v8.0.0

The problem with this is that I can't merge it back to the default-branch yet, as I don't have a new major release planned yet (due to the breaking changes). While I do intend to release a new major release in the future (which also will include major upgrades of other dependencies), I don't know yet when that is going to happen, because of the workload of my current job.

So it may take a while to integrate this.

The remaining Todos for this PR, as far as I can tell, are:

- [ ] Change the reference value for the failing test from `true` to `false` (which is ok, if it's part of the next major release)
- [ ] Move the new data-type `date-time` to a separate PR
- [ ] Deal with unknown used formats
- [ ] Restore everything that was in the `dist/`-directory from the base. That directory contains only generated files and these files should not be committed, because they are only newly generated on every release (as they are prone to merge conflicts, otherwise).
- [ ] Add dedicated test-cases for this new feature

That said, I **really appreciate** your effort and would like to see this feature in a future version as I think that it's a valuable new feature.

**Comment by gigaga** [comment]
Hi @codekie ,
I close this PR, I'm going to create another from my "feature/discriminator" branch.
Regards,

**Pull request closed without merging**